src/hotspot/cpu/aarch64/assembler_aarch64.hpp

```diff
@@ -2860,6 +2867,29 @@ template<typename R, typename... Rx>
   // FMULX - Vector - Scalar
   INSN(fmulxvs, 1, 0b1001);
 
+#undef INSN
+
+#define INSN(NAME, op1, op2)                                                                        \
+  void NAME(FloatRegister Vd, SIMD_Arrangement T, FloatRegister Vn, FloatRegister Vm, int index) {  \
+    starti;                                                                                         \
+    assert(T == T4H || T == T8H || T == T2S || T == T4S, "invalid arrangement");                    \
+    assert(index >= 0 &&                                                                            \
+               ((T == T2S && index <= 1) || (T != T2S && index <= 3) || (T == T8H && index <= 7)),  \
+           "invalid index");                                                                        \
+    assert((T != T4H && T != T8H) || Vm->encoding() < 16, "invalid source SIMD&FP register");       \
+    f(0, 31), f((int)T & 1, 30), f(op1, 29);                                                        \
+    f(0b01111, 28, 24);                                                                             \
+    if (T == T4H || T == T8H) {                                                                     \
+      f(0b01, 23, 22), f(index & 0b11, 21, 20), lrf(Vm, 16), f(op2, 15, 12), f(index >> 2 & 1, 11); \
+    } else {                                                                                        \
+      f(0b10, 23, 22), f(index & 1, 21), rf(Vm, 16), f(op2, 15, 12), f(index >> 1, 11);             \
+    }                                                                                               \
+    f(0, 10), rf(Vn, 5), rf(Vd, 0);                                                                 \
```

Comment:
```suggestion
#define INSN(NAME, op1, op2)                                                              \
  void NAME(FloatRegister Vd, SIMD_Arrangement T, FloatRegister Vn, FloatRegister Vm, int index) { \
    starti;                                                                               \
    assert(T == T4H || T == T8H || T == T2S || T == T4S, "invalid arrangement");          \
    assert(index >= 0 &&                                                                  \
               ((T == T2S && index <= 1) || (T != T2S && index <= 3) || (T == T8H && index <= 7)), \
           "invalid index");                                                              \
    assert((T != T4H && T != T8H) || Vm->encoding() < 16, "invalid source SIMD&FP register"); \
    f(0, 31), f((int)T & 1, 30), f(op1, 29); f(0b01111, 28, 24); f(0b01, 23, 22);         \
    if (T == T4H || T == T8H) {                                                           \
      f(index & 0b11, 21, 20), lrf(Vm, 16);                                               \
    } else {                                                                              \
      f(index & 1, 21), rf(Vm, 16);                                                       \
    }                                                                                     \
    f(op2, 15, 12), f(index >> 1, 11); f(0, 10), rf(Vn, 5), rf(Vd, 0);                    \
```
I think it's a bit easier to see what's going on here if we lose the duplicated code.

Reply:
Looks like that's incorrect: the 22th-23th bits and 11th bits differ.

Reply:
It's untested. What I'm trying to say is that we shouldn't duplicate stuff. Perhaps I should have been clearer.

Separate the fields into what is the same, and what is different. Put the different inside the if. Put the common outside the if.

Reply:
The only matching field out of the sequence of four under the if/else is `f(op2, 15, 12)`, but OK, I can hoist it if this makes things clearer.

Reply:
Fixed by https://github.com/openjdk/jdk/pull/18487/commits/a28bbcd3bbe7555e715b107b71c18efaf72a7ce7.